README.md
## yf-fpm-server

### CHANGELOG

点击查看更新日志 [CHANGELOG](CHANGELOG.md)

### ROADMAP

点击查看 [ROADMAP](ROADMAP.md)

### 0.OVERVIEW
> yf-fpm-server是一款轻量级的api服务端，可通过插件集成数据库(mysql,mongodb)的数据操作，灵活扩展自定义业务逻辑

* 源码地址: https://github.com/team4yf/yf-fpm-server.git
* 微内核，源码
* 基于koa2框架
* 支持key + secret安全验证
* 支持接口权限验证
* 支持hook钩子扩展
* 支持接口多版本同时在线

#### 设计概要

##### 　背景
> 团队的产品从单平台，慢慢扩展到多个客户端，且有多个异构系统的数据交互，所以需要一个统一的数据输入输出口。

- Node 作为一个事件驱动的V8服务端语言，非常适合处理高并发，非事务密集型的场景。
- 使用koa2作为http框架，因为它够轻，够简洁，几乎没有学习成本，而且体积小，自然坑也少 [偷笑]。
- 没有使用restful风格，因为业务需要对数据有权限限制，而且业务交集很多，对路由的管理成本就大。
- 采用taobao和jd的开放平台的设计方案，定义统一的入口，通过参数定位业务接口，实现灵活的业务开发。

##### 　定义

　　　　服务只有一个入口 **/api** ; 只接受application/json方式的请求,定义如下

* 传入参数结构

| 参数名 | 类型 | 是否必须 | 参数说明 | 默认值 | 示例 |
|:-------:|:-------:|:-----:|:-----:|:-----:|:-----:|
| method | String | Y | 需要调用的业务函数 | 无 | |
| appkey | String | Y | 应用被分配的密钥 | 无 | |
| timestamp | Number | Y | 应用端的时间戳，用于验证请求的时效性 | 无 | 13位时间戳 |
| v | String | N | 调用的服务端接口的版本号 | 无 | |
| param | Object | N | 业务函数需要用到的参数，以JsonObject的形式传入 | 无 | |
| sign | String | Y | 将接口参数进行升序排列，如 appkey,method,param,timestamp添加一个masterKey=xxx[此处的key来自于注册的key] 组合成appkey=123&masterKey=xxx&param=44444&tmiestamp=140932932932[所有的参数值使用urlencode] 过md5加密，生成一个32位的密钥 | 无 | |

* 输出参数结构

| 参数名 | 类型 | 是否必须 | 参数说明 |
|:-------:|:-------:|:-----:|:-----:|
| errno | Number | Y | 业务函数的错误代码，通常为0，表示正常执行，<0 则表示执行错误，可通过应用说明获取到具体的错误原因。|
| message | String | N | 通常在执行出错的情况下，会输出错误的信息。 |
| timestamp | Number | N | 返回服务端处理完信息之后的时间戳。 |
| data | Object/Array	 | Y |  一般的查询类的业务函数，会在该字段下携带查询结果信息；具体是Object类型还是Array类型则根据不同的业务函数的说明而定。 |
| error | Object | N | 	错误信息的详细内容 |

　　这样的设计不能满足现在的restful范式，但是能满足小规模团队的需求，可提高业务实现的开发效率；

### 1.Install
`
$ npm install yf-fpm-server --save
`

### 2.Code
`
$ vi app.js
`
```javascript
'use strict';
const { Fpm } = require('yf-fpm-server');
const fpm = new Fpm();
const biz = fpm.createBiz('0.0.1');
biz.addSubModules('test', {
  foo: async (args, ctx, before) => {
    return Promise.reject({errno: -3001})
  }
});
fpm.addBizModules(biz);
fpm.run()
  .then(() => {
    fpm.logger.info('ready ...')
  });
```
### 3.Run It
`
$ node app.js
`

### 4.Global Error Define

#### System 系统错误定义

| 异常编号 | 异常代码 | 异常信息 | 排查办法 |
|:-------:|:-------:|:-----:|:-----:|
| -900 | LOST_PARAM | param: [${col}] required! | 检查接口传入的参数 |
| -901 | NO_POST_DATA | post data is empty! | 接口未传入参数 |
| -902 | TIMEZONE_OVER | your time zone not sync the server! | 传入参数中的时间戳与服务器的事件戳相差超过了30分钟 |
| -903 | SIGN_ERROR | param sign error! | 参数签名错误，请检查参数拼接过程是否异常 |
| -904 | PLUGIN_LOAD_ERROR | missing plugin ! plugin: ${ pname } dependent plugin: ${ dname } | 某个插件缺失了另外一个插件的依赖 |
| -905 | PARAM_IS_NOT_JSON | Param is not json! | 参数Param不是一个正确的json格式 |
| -906 | SQL_INJECTION | you have sql keyword! ex:['drop ','delete ','truncate ',';','insert ','update ','set ','use '] | 传入的参数可能有sql注入 |
| -907 | NOT_LATEST | Not the latest version | 不是最新的版本，请及时更新 |
| -908 | NOT_METHOD | Cant find the method! | 调用的业务函数没有定义 |
| -909 | BIZ_MODULE_EXTEND_ERROR | Biz Version ${version} Not Exists! | 扩展业务的接口版本不存在 |
| -910 | TABLE_REQUIRED | table required! | 使用 DB 插件时，需要传入 table 参数 |
| -911 | VERSION_UNDEFINED | version not defined! | 调用的业务函数版本不存在 |
| -912 | AUTH_ERROR | auth error! plz check your appkey ~ ' | 客户端的appkey未授权 |
| -913 | ROOT_ERROR | auth error! plz check roots of your app  ~  | 客户端的appkey权限不够 |
| -914 | BIZ_VERSION_REQUIRED | Biz Version Required | 业务组件版本未定义 |
| -915 | HOOK_VERSION_REQUIRED | hook version cant be undefined | 钩子版本未定义 |
| -916 | HOOK_HANDLER_REQUIRED | hook handler cant be undefined | 钩子函数未定义 |
| -920 | UNDEFINED_EXCEPTION | UnDefinedException | 未定义的异常 |
| -999 | UNCAUGHT_ERROR | System uncaughtException | 系统异常 |

CHANGELOG.md
# 2.7.0 (2021-05-19)
Modify:
- `Logger`
  
  Make it more flexiable, eg: define where the log output!

# 2.6.2 (2019-04-22)
Add:
- `SSL`
  - command:

    ```bash
    $ openssl genrsa -out privatekey.pem 1024
    $ openssl req -new -key privatekey.pem -out certrequest.csr 
    $ openssl x509 -req -in certrequest.csr -signkey privatekey.pem -out certificate.pem
    ```
    
# 2.5.3 (2019-03-25)
Add:
- Extend the `config.*.json` to the config Object, the `*` is the `NODE_ENV` of the `process.ENV`; 
  - Such as `process.ENV.NODE_ENV === 'DEV'` the `config.dev.json` file will be loaded and assign to the config object.
  - The `*` must be lowercase.
  - The `*` default value is `DEV`.


# 2.5.1 (2019-03-04)

Add:
- Add `disableBodyParser` config.
  
  use `new Fpm({ disableBodyParser: ['/notify']})` to disable the `/notify` post handler

  add a middleware:
  ```javascript
    app.use(async(ctx, next) => {
      delete ctx.disableBodyParser;
      const disableBodyParser = this.get('disableBodyParser');
      if(disableBodyParser){
        let urls;
        if(_.isArray(disableBodyParser)){
          urls = disableBodyParser;
        }else if(_.isString(disableBodyParser)){
          urls = disableBodyParser.split(',');
        }
        const urlPath = ctx.request.url ;
        _.map(urls, url => {
          if(_.endsWith(urlPath, url)){
            ctx.disableBodyParser = true;
          }
        })
      }
      await next()
    })
  ```
- Support https

Simple Config:
```javascript
{
  ssl: {
    key: 'ssl/server.key',  // the key filepath
    cert: 'ssl/server.crt', // the crt filepath
    port: 9443,             // the https port
  },
}
```

How to make a ssl key/cert?

[Here](https://www.cnblogs.com/liyulong1982/p/6106129.html)

# 2.5.0 (2018-11-12)

Add:
- `package-lock.json`
- `fpm.getEnv(envKey, defaultValue)` get the process.env key-values

Remove:
- yarn.lock

# 2.4.12 (2018-06-08) 

Add:
- `fpm.getPluginRef(pname)` get the origin ref of the plugin by the plugin name, ex: `getPluginRef('mysql')` or `getPluginRef('fpm-plugin-mysql')`

- `fpm._webhook_events` record the webhook events
  contains: { topic!String, message!Object, at!timestamp}

- Use `curl` to perform webhook
  Webhook Url: `/webhook/:upstream/:type/:data`
  Publish Topic: `#webhook/${upstream}/${type}`
  `$ curl -H "Content-Type:application/json" -X POST --data '{"id":10}' http://localhost:9999/webhook/test/foo/bar `

# 2.4.11 (2018-06-07)
BugFix:
- `publish(topic, data)`

# 2.4.10 (2018-06-07)
Modify:
- `async execute(method, args, v, ctx)` The `v` Change To Be Optional
- add `fpm._publish_topics`, it save all the topics published by the biz functions
- `publish(topic, data)` , It May Throw An Exception When The TOPIC Published By Other Function

# 2.4.9 (2018-05-31)
BugFix:
- masterkey lost from 
  before: `let sign = (args) => {` ; after: `let sign = (args, apps) => {`
  

# 2.4.7 (2018-05-31)
Add:
- `execShell(shellPath:String!, params:[]?) => Promise`
  Add an execShell method For Run `.sh` or `.bat` files. 
 
  *The ShellPath Should Be An Abstract Path*
BugFix:
- `Digest already called` AT File `yf-fpm-server/lib/utils/kit.js:25:7`
  https://github.com/nodejs/node-v0.x-archive/issues/1415

# 2.4.2 (2018-05-28)
Update:
- `loadPlugin`
  - Add Field `package` For Each Plugin, It links this origin reference of the plugin module
  - Add Field `isBound` For Each Plugin, It means plugin invoke `bind()` ok

Files:
- Move `loadPlugin` from `bin\app.js` to `utils\plugin.js`

# 2.4.1 (2018-05-26)
Add 3 Field For Plugin:
- info: `packageInfo` 
  The package.json For The Plugin
- npm:`https://www.npmjs.com/package/${packageInfo.name}`
  The HomePage For The Plugin In `npmjs.com`
- registry: `http://registry.npmjs.org/${packageInfo.name}` 
  The Registry Info Of The Plugin

Remove:
- remove `deletedir`

Update:
- format all error code & errno

# 2.4.0 (2018-05-25)
update:
- koa@2.5.x

# 2.3.1 (2018-05-25)
Add:
- default config `server.hostname` & `server.domain`
  ```javascript
  server:{
    hostname: '0.0.0.0',
    domain: 'localhost',
    port: 9999
  }
  ```
- Return A Promise In `fpm.run()`, it will be called after the server startup
  ```javascript
  app.run()
    .then(fpm => {
      fpm.logger.info('ready ...')
    })
  ```

# 2.3.0 (2018-05-19)

ADD:
- fpm._counter : count the api calls
- fpm._project_info: the project package.json

Exception Code: from `900` ~ `999`

Update:
- yf-fpm-nodejs-client@1.0.15

Remove:
- gulp
- jquery
- materialize-css
- log
- moment
- md5
- ncp
- koa-multer

# 2.2.25 (2018-05-14)

Remove: 
- views/
- public/ 
- admin.js

TODO:
- implement a fpm-plugin-admin

# 2.2.24 (2018-05-13)

ADD:
- Add Action Before Api Invoke
`ctx.fpm.runAction('CALL_API', ctx.fpm, ctx, { method, param, v })`
- Attach An Argument `ctx` At Function `core(method, args, v, fpm, ctx)`
- Attach An Argument `ctx` At Each Handler `handler(args, ctx, [])`
  - The Last Argument should be replaced by the output of before_*_hook, It always be an Array
- Attach An Argument `ctx` At `hook.runHook('before_' + method, args, v, ctx)`


# 2.2.24 (2018-04-28)

TODO:
- Add Ip Filter In File `permission.js`ss

# 2.2.24 (2018-04-25)

Remove

- The `BEFORE_MODULES_ADDED` & `AFTER_MODULES_ADDED` Action Hook Points May Invoke Multi Times; `fpm.addBizModules(biz:Biz)` can be called many times

- In The Middleware `response.js`;  Support More Type Of Result Not Only `Object`
  ```javascript
  ctx.success = (result, msg) => {
    data.message = msg || ''
    if(_.isPlainObject(result)){
      if(_.has(result, data)){
        data.data = result.data
        return
      }
    }    
    data.data = result   
  }
  ```

Add

- `extendModule(name:String!, module:Object/Function!, version:String?)` for plugin in `DEV` mode
- `getPlugins()` show all loaded plugins
- `isPluginInstalled(name:String!)`

Remove 

- example\runner.js 

# 2.2.21 (2017-07-15)

Add

- Fpm
  - Admin
    - login

# 2.2.5 (2017-07-01)

Add

- Fpm
  - `createRouter`

src/error.js
export default {
  System: {
    LOST_PARAM: col => { 
      return {
        errno: -900,
        code: 'LOST_PARAM',
        message: `param: [${col}] required!`
      }
    },
    NO_POST_DATA: { 
      errno: -901, 
      code: 'NO_POST_DATA', 
      message: 'post data is empty!' 
    },
    TIMEZONE_OVER: { 
      errno: -902, 
      code: 'TIMEZONE_OVER', 
      message: 'your time zone not sync the server!'
    },
    SIGN_ERROR: {
      errno: -903,
      code: 'SIGN_ERROR',
      message: 'param sign error!'
    },
    PLUGIN_LOAD_ERROR: (pname, dname) =>{
      return {
        code: 'PLUGIN_LOAD_ERROR',
        errno: -904,
        message: `missing plugin ! plugin: ${ pname } dependent plugin: ${ dname }`
      }
    },
    PARAM_IS_NOT_JSON: {
      errno: -905,
      code: 'PARAM_IS_NOT_JSON',
      message: 'Param is not json!'
    },

    SQL_INJECTION:{
      errno: -906,
      code: 'SQL_INJECTION',
      message: `you have sql keyword! ex:['drop ','delete ','truncate ',';','insert ','update ','set ','use ']`
    },
    NOT_LATEST:{
      errno: -907,
      code: 'NOT_LATEST',
      message: 'Not the latest version'
    },
    NOT_METHOD:{
      errno: -908,
      code: 'NOT_METHOD',
      message: 'Cant find the method!'
    },
    BIZ_VERSION_NOT_FNOUND: (version) =>{
      return {
        code: 'BIZ_MODULE_EXTEND_ERROR',
        errno: -909,
        message: `Biz Version ${version} Not Exists!`
      }
    },
    TABLE_REQUIRED:{
      errno: -910,
      code: 'TABLE_REQUIRED',
      message: 'table required!'
    },
    VERSION_UNDEFINED:{
      errno: -911,
      code: 'VERSION_UNDEFINED',
      message: 'version not defined!'
    },
    //uncaught
    AUTH_ERROR:{
      errno: -912,
      code: 'AUTH_ERROR',
      message: 'auth error! plz check your appkey ~ '
    },
    ROOT_ERROR:{
      errno: -913,
      code: 'ROOT_ERROR',
      message: 'auth error! plz check roots of your app  ~ '
    },
    BIZ_VERSION_REQUIRED: {
      code: 'Biz-Version-Required',
      errno: -914,
      message: 'Biz Version Required'
    },
    HOOK_VERSION_REQUIRED: {
      errno: -915,
      code: 'Hook-Version-Required',
      message: 'hook version cant be undefined'
    },
    HOOK_HANDLER_REQUIRED: {
      errno: -916,
      code: 'Hook-Handler-Required',
      message: 'hook handler cant be undefined'
    },
    ONLY_POST_ALLOWED: {
      errno: -917,
      code: 'ONLY_POST_ALLOWED',
      message: 'only post method allowed~'
    },
    TOPIC_BEEN_PUBLISHED: {
      errno: -918,
      code: 'TOPIC_BEEN_PUBLISHED',
      message: 'topic has been published by other biz'
    },
    UNDEFINED_EXCEPTION: {
      errno: -920,
      code: 'UnDefinedException',
      message: 'UnDefinedException',
    },

    UNCAUGHT_ERROR:{
      errno: -999,
      code: 'UNCAUGHT_ERROR',
      message: 'uncaughtException!'
    },

  }
}
